src/smtencoding/FStar.SMTEncoding.Solver.fs

```diff
@@ -651,34 +657,141 @@ let ask_and_report_errors env all_labels prefix query suffix =
         @ max_fuel_max_ifuel
     in
 
-    let check_one_config config (k:z3result -> unit) : unit =
+    let check_one_config config : z3result =
           if Options.z3_refresh() then Z3.refresh();
           Z3.ask config.query_range
                   (filter_assertions config.query_env config.query_hint)
                   config.query_hash
                   config.query_all_labels
                   (with_fuel_and_diagnostics config [])
                   (Some (Z3.mk_fresh_scope()))
-                  k
                   (used_hint config)
     in
 
-    let check_all_configs configs =
-        let report errs = report_errors ({default_settings with query_errors=errs}) in
-        fold_queries configs check_one_config process_result report
+    let check_all_configs configs : either<list<errors>, query_settings> =
+        fold_queries configs check_one_config process_result
+    in
+
+    let quake_and_check_all_configs configs =
+        let lo   = Options.quake_lo () in
+        let hi   = Options.quake_hi () in
+        let seed = Options.z3_seed () in
+        let name = full_query_id default_settings in
+        let quaking = hi > 1 in
+        let hi = if hi < 1 then 1 else hi in
+        let lo =
+            if lo < 1 then 1
+            else if lo > hi then hi
+            else lo
+        in
+        let run_one (seed:int) =
+            (* Here's something annoying regarding --quake:
+             *
+             * In normal circumstances, we can just run the query again and get
+             * a slightly different behaviour because of Z3 accumulating some
+             * internal state that doesn't get erased on a (pop). So we simply repeat
+             * the query then.
+             *
+             * But, if we're doing --z3refresh, we will always get the exact
+             * same behaviour by doing that, so we do want to set the seed in this case.
+             *
+             * Why not always set it? Because it requires restarting the solver, which
+             * takes a long time.
+             *
+             * Why not use the (set-option smt.random_seed ..) command? Because
+             * it seems to have no effect just before a (check-sat), so it needs to be
+             * set early, which basically implies restarting.
+             *
+             * So we do this horrendous thing.
+             *)
+            if Options.z3_refresh ()
```

Comment:
Agreed that looks a bit hacky. Another option might be to ignore `--z3_refresh` (with a console message saying that quaking does not restart z3) for the purposes of quaking.

Reply:
I think I'd rather keep the hack than ignore --z3refresh. A funny thing I found is that for this:
```fstar
val lem (a b c : int) : Lemma ((a + b) * c = a * c + b * c)
let lem a b c = ()
```
I get 902/1000 successes normally, but 998/1000 with refresh. I can't really explain that discrepancy, but we might want to investigate these things in the future.